rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+  - As a rule of thumb: Expressions which can be used as list items (without parentheses)
+- **Absorption:**
+  A multiline expression can have an alternative layout depending on the context.
+  In that case, it will start on the end of the current line instead of a new line,
+  potentially saving a level of indentation of its content.
+  ```nix
+  {
+    # The right-hand side of bindings is an example of a situation where absorption improves the style.
+    absorbed = with bar; [
+      1
+      2
+      3
+    ];
+    notAbsorbed =
+      with bar; # Placing a comment here will force the non-absorbed, multiline layout.
+      [
+        1
+        2
+        3
+      ];
+
+    # In this case, absorption does not reduce the indentation level of the set.
+    absorbed' =
+      let
+        qux = 1;
+      in
+      # { is absorbed
+      bar: baz: {
+        # <-- same level
+      };
+
+    notAbsorbed' =
+      let
+        qux = 1;
+      in
+      way:
+      too:
+      many:
+      arguments:
+      {
+        # <-- same level
+      };
+  }
+  ```
+- **Absorbable Terms:**
+  Attribute sets, lists, and multiline `''` strings are called absorbable terms. Parenthesized absorbable terms are absorbable terms again too.
+
+## Initial standard Nix format
+
+- The formatter should be as "pure" as possible, i.e. different input formats of the "same" code (same AST with comments) should result in the same output format.
+  - The formatter may take the input formatting into account in some cases in order to preserve multi-line syntax elements (which would otherwise have been contracted by the rules).
+- Line breaks may be added or removed, but empty lines must not be created. Single empty lines must be preserved, and consecutive empty lines must be collapsed into a single empty line.
+  This allows the formatter to expand or compact multi-line expressions, while still allowing grouping of code.
+
+  For example, formatting this code:
+  ```nix
+  [
+    0 10
+
+    (
+      20 + 1
+    )
+
+
+    30
+  ]
+  ```
+
+  turns into this:
+  ```nix
+  [
+    0  # Line break added
+    10
+
+    (20 + 1) # Line breaks removed
+             # Consecutive empty lines turned into a single empty line
+    30
+  ]
+  ```
+
+- Expressions of the same kind that can be treated as a sequence of expressions on the same level should be treated as such, even though they are technically parsed as a nested tree.
+  - This applies to else-if chains, functions with multiple arguments, some operators, etc.
+  - Example:
+    ```nix
+    # This is treated as a sequence of if-then-else chains, instead of indenting the second if as part of the else body
+    if cond1 then
+      foo
+    else if cond2 then
+      bar
+    else
+      baz
+    ```
+
+- Indentation should reflect the expression structure.
+  Example:
+  ```nix
+  # Bad, the indentation misleads the user
+  {
+    foo = {
+    bar = if
```

Comment:
This would violate these specifications of the bindings section:

> The semicolon in bindings must always be placed on the same line as the expression it concludes.

The semicolon must be behind the `20`.

> Within bindings, if the first and last line are not indented, the absorbed style is used, otherwise newline and indent

Your example uses absorbed style while the last line with an expression (`20;`) is indented.

What this rule together with the single indentation level growth rule effectively prevents is the placement of structural element like `else` or `in` on the top-level indent of the keys. See how your `else` is on the same indent as `bar`?

It's a bit unintuitive, so perhaps this "don't encroach on the keys indent" rule could be added as an explicit guideline.

Reply:
We now improved the clarity of the bindings rule and added your example with https://github.com/NixOS/rfcs/pull/166/commits/685ee9bf80abdfaceffe1e423f9c0a0609b82890 :)

Reply:
@infinisil I would like you to clarify: Is the effect that attribute names have an exclusive indent level intended?

If so, could that be added as an explicit rule? The rules of the RFC appear to indirectly cause this effect but it's not obvious that they do or whether that's the reason for their existence. This could make some of them feel less arbitrary.